File: DataQuality/DataQualityTools/src/DQTBackgroundMon.cxx
/*
  Copyright (C) 2002-2019 CERN for the benefit of the ATLAS collaboration
*/

#include "DataQualityTools/DQTBackgroundMon.h"

DQTBackgroundMon::DQTBackgroundMon( const std::string& name, ISvcLocator* pSvcLocator )
: AthMonitorAlgorithm(name,pSvcLocator)
{}

DQTBackgroundMon::~DQTBackgroundMon() {}

StatusCode DQTBackgroundMon::initialize() {
    ATH_CHECK( m_RawInfoSummaryForTagKey.initialize() );
    ATH_CHECK( m_LArCollisionTimeKey.initialize() );
    ATH_CHECK( m_MBTSCollisionTimeKey.initialize() );
    ATH_CHECK( m_TileCellContainerKey.initialize() );
    ATH_CHECK( m_LUCID_RawDataContainerKey.initialize() );
    ATH_CHECK( m_BeamBackgroundDataKey.initialize() );
    ATH_CHECK( m_VertexContainerKey.initialize() );

    return AthMonitorAlgorithm::initialize();
}

StatusCode DQTBackgroundMon::fillHistograms( const EventContext& ctx ) const {
    ATH_MSG_DEBUG("In DQTBackgroundMon::fillHistograms()");
    using namespace Monitored;

    auto group = getGroup("default");

    auto filled = Scalar<bool>("filled",false);
    auto empty = Scalar<bool>("empty",false);
    auto unpairIso = Scalar<bool>("unpairIso",false);
    auto unpairNonIso = Scalar<bool>("unpairNonIso",false);

    RH<xAOD::EventInfo> eventInfo{GetEventInfo(ctx)};
    RH<RawInfoSummaryForTag> rawInfo(m_RawInfoSummaryForTagKey,ctx);
    RH<LArCollisionTime> tps(m_LArCollisionTimeKey,ctx);
    RH<MBTSCollisionTime> mbtsTime(m_MBTSCollisionTimeKey,ctx);
    RH<TileCellContainer> tileCellContainer(m_TileCellContainerKey,ctx);
    RH<LUCID_RawDataContainer> LUCID_RawDataContainer(m_LUCID_RawDataContainerKey,ctx);
    RH<BeamBackgroundData> beamBackgroundData(m_BeamBackgroundDataKey,ctx);
    RH<xAOD::VertexContainer> vertexContainer(m_VertexContainerKey,ctx);

    if ( eventInfo.isValid() ) {
        if (!getTrigDecisionTool().empty()) {
            unsigned int bgCode = getTrigDecisionTool()->getBGCode();
            filled = bgCode & (1<<m_filledBG);
            empty = bgCode & (1<<m_emptyBG);
            unpairIso = bgCode & (1<<m_unpairIsoBG);
            unpairNonIso = bgCode & (1<<m_unpairNonIsoBG);
        }

        std::vector<int> bitVec = {};
        std::vector<int> filledVec = {};
        std::vector<int> emptyVec = {};
        std::vector<int> unpairIsoVec = {};
        std::vector<int> unpairNonIsoVec = {};

        for ( int bkg=0; bkg<m_nBkgWords; bkg++ ) {
            unsigned int bgWord = eventInfo->eventFlags(xAOD::EventInfo::Background);
            ATH_MSG_DEBUG("Background word is "<<bgWord);
            if ( !(bgWord & 1<<bkg) ) {
                continue;
            }

            bitVec.push_back(bkg);
            if (filled) {
                filledVec.push_back(bkg);
            } else if (empty) {
                emptyVec.push_back(bkg);
            } else if (unpairIso) {
                unpairIsoVec.push_back(bkg);
            } else if (unpairNonIso) {
                unpairNonIsoVec.push_back(bkg);
            }
        }

        auto bitSet = Collection("bitSet",bitVec);
        auto bitSetFilled = Collection("bitSetFilled",filledVec);
        auto bitSetEmpty = Collection("bitSetEmpty",emptyVec);
        auto bitSetUnpairIso = Collection("bitSetUnpairIso",unpairIsoVec);
        auto bitSetUnpairNonIso = Collection("bitSetUnpairNonIso",unpairNonIsoVec);
        fill(group,bitSet,bitSetFilled,bitSetEmpty,bitSetUnpairIso,bitSetUnpairNonIso);
    } else {
        ATH_MSG_WARNING("Event data invalid. Background word histograms are being skipped.");
    }

    if ( rawInfo.isValid() ) {
        auto nPixSPs = Scalar<float>("nPixSPs",rawInfo->getNpixSPs());
        auto nSctSPs = Scalar<float>("nSctSPs",rawInfo->getNsctSPs());
        fill(group,nPixSPs,nSctSPs,unpairIso,unpairNonIso);
        if ( nPixSPs<m_upPixSP )
            fill("lowMultiplicityPixGroup",nPixSPs,unpairIso,unpairNonIso);
        if ( nSctSPs<m_upSctSP )
            fill("lowMultiplicitySctGroup",nSctSPs,unpairIso,unpairNonIso);
    } else {
        ATH_MSG_WARNING("Raw data invalid. nPix/nSct is being skipped.");
    }

    if ( tps.isValid() ) {
        if ( tps->ncellA()>m_LArECSideCut && tps->ncellC()>m_LArECSideCut ) {
            auto LArECTimeDiff = Scalar<float>("LArECTimeDiff", tps->timeA()-tps->timeC());
            fill(group,LArECTimeDiff);
        }
    } else {
        ATH_MSG_WARNING("TPS data invalid. LArECTimeDiff is being skipped.");
    }

    if ( mbtsTime.isValid() ) {
        if ( mbtsTime->ncellA()>m_MBTSSideCut && mbtsTime->ncellC()>m_MBTSSideCut ) {
            auto MBTStimeDiff = Scalar<float>("MBTStimeDiff", mbtsTime->time());
            fill(group,MBTStimeDiff);
        }
    } else {
        ATH_MSG_WARNING("MBTSCollisionTime data invalid. MBTStimeDiff is being skipped.");
    }

    if ( tileCellContainer.isValid() ) {
        auto MBTSCount = Scalar<int>("MBTSvetoHits",0);
        for ( auto tileCell : *tileCellContainer ) {
            if (tileCell->energy()>m_MBTSThresholdCut &&
                (tileCell->qbit1() & m_MBTSMask)==m_MBTSPattern &&
                fabs(tileCell->time())<m_MBTSTimeCut )
                MBTSCount++;
        }
        fill(group,MBTSCount);
    } else {
        ATH_MSG_WARNING("TileCell data invalid. MBTSvetoHits is being skipped.");
    }

    if ( LUCID_RawDataContainer.isValid() ) {
        auto LUCIDCounter = Scalar<int>("LUCIDCounter",0);
        for ( auto LUCID_RawData : *LUCID_RawDataContainer ) {
            LUCIDCounter += LUCID_RawData->getNhitsPMTsideA() + LUCID_RawData->getNhitsPMTsideC();
        }
        fill(group,LUCIDCounter);
    } else {
        ATH_MSG_WARNING("LUCID data invalid. LUCID is being skipped.");
    }

    if ( m_doMuons ) {
    	if (beamBackgroundData.isValid() ) {
            // Muon segments and clusters
            std::vector<double> muSegXVec = {};
            std::vector<double> muSegYVec = {};
            for (int iSeg=0; iSeg<beamBackgroundData->GetNumSegment(); iSeg++) {
                const Trk::Segment* trkSeg = beamBackgroundData->GetIndexSeg(iSeg);
                const xAOD::MuonSegment* muonSeg = dynamic_cast<const xAOD::MuonSegment*>(trkSeg);
                if (muonSeg) {
                    muSegXVec.push_back(muonSeg->x()*1e-3);
                    muSegYVec.push_back(muonSeg->y()*1e-3);
                }
            }
            auto muSegXCollection = Collection("muonSegmentX",muSegXVec);
            auto muSegYCollection = Collection("muonSegmentY",muSegYVec);

            std::vector<double> clusterEnergyVec = {};
            std::vector<double> clusterEtaVec = {};
            std::vector<double> clusterPhiVec = {};
            std::vector<double> clusterTimeVec = {};
            for (int i=0; i<beamBackgroundData->GetNumMatched(); i++) {
                const xAOD::CaloCluster* cluster = beamBackgroundData->GetIndexClus(i);
                clusterEnergyVec.push_back(cluster->e()*1e-3);
                clusterEtaVec.push_back(cluster->eta());
                clusterPhiVec.push_back(cluster->phi());
                clusterTimeVec.push_back(cluster->time());
            }
            auto clusterEnergyCollection = Collection("clusterEnergy",clusterEnergyVec);
            auto clusterEtaCollection = Collection("clusterEta",clusterEtaVec);
            auto clusterPhiCollection = Collection("clusterPhi",clusterPhiVec);
            auto clusterTimeCollection = Collection("clusterTime",clusterTimeVec);

            fill(group,muSegXCollection,muSegYCollection,clusterEnergyCollection,
                 clusterEtaCollection,clusterPhiCollection,clusterTimeCollection);

            // fake jets
            if ( beamBackgroundData->GetNumNoTimeTight() ) {
                std::vector<int> jetIndexVec = {};
                for (int i=0; i<beamBackgroundData->GetNumJet(); i++) {
                    const xAOD::Jet* jet = beamBackgroundData->GetIndexJet(i);
                    int jetIndex = beamBackgroundData->GetIndexJet(jet);
                    jetIndexVec.push_back(jetIndex);

                    // Leading jet
                    if ( jetIndex==0 ) {
                        auto pt = Scalar<double>("leadingJetPt", jet->pt());
                        auto eta = Scalar<double>("leadingJetEta", jet->eta());
                        auto phi = Scalar<double>("leadingJetPhi", jet->phi());
                        double timeValue = jet->getAttribute<float>(xAOD::JetAttribute::Timing);
                        auto time = Scalar<double>("leadingJetTime", timeValue);
                        double emfValue = jet->getAttribute<float>(xAOD::JetAttribute::EMFrac);
                        auto emf = Scalar<double>("leadingJetEMFrac", emfValue);

                        std::vector<float> sumPt_v;
                        auto trkPtAttr = xAOD::JetAttribute::SumPtTrkPt1000;
                        bool hasSumPt = jet->getAttribute<std::vector<float>>(trkPtAttr, sumPt_v);
                        double sumPtTrk = hasSumPt ? sumPt_v[0] : 0.;
                        auto chf = Scalar<double>("leadingJetCHF", sumPtTrk / pt);

                        fill(group,pt,eta,phi,time,emf,chf);
                    }
                }
                auto jetIndexCollection = Collection("jetIndex",jetIndexVec);

                auto bcid = Scalar<int>("bcid",2);
                auto isAC = Scalar<bool>("isAC", beamBackgroundData->GetDirection()>0);
                auto isCA = Scalar<bool>("isCA", beamBackgroundData->GetDirection()<0);
                bool isOneSided = beamBackgroundData->GetNumOneSidedLoose();
                auto isACOneSided = Scalar<bool>("isACOneSided", isAC && isOneSided );
                auto isCAOneSided = Scalar<bool>("isCAOneSided", isCA && isOneSided );
                fill(group,jetIndexCollection,bcid,isAC,isCA,isACOneSided,isCAOneSided);
            } // end fake jets
        } else {
            ATH_MSG_WARNING("BeamBackgroundData data invalid. Muons are being skipped.");
        }
    } // end muons

    if (vertexContainer.isValid()) {
        // count vertices, excluding the dummy vertex
        auto nVertex = Scalar<int>("nVertex", vertexContainer->size()-1);
        fill(group,nVertex,unpairIso,unpairNonIso);
    } else {
        ATH_MSG_DEBUG("Primary vertex data invalid. Primary vertex counts are being skipped.");
    }
    return StatusCode::SUCCESS;
}

File: DataQuality/DataQualityTools/python/DQTBackgroundMon.py
#
#  Copyright (C) 2002-2020 CERN for the benefit of the ATLAS collaboration
#

def DQTBackgroundMonAlgConfig(flags, isOld=False):
    if isOld: # replace with helper for the old configuration
        from AthenaMonitoring import AthMonitorCfgHelperOld as AthMonitorCfgHelper
        from .DataQualityToolsConf import DQTBackgroundMon
    else:
        from AthenaMonitoring import AthMonitorCfgHelper
        from AthenaConfiguration.ComponentFactory import CompFactory
        DQTBackgroundMon = CompFactory.DQTBackgroundMon
    helper = AthMonitorCfgHelper(flags, 'DQTBackgroundMonAlgCfg')
    

    monAlg = helper.addAlgorithm(DQTBackgroundMon,'DQTBackgroundMonAlg')
    monAlg.Muons = True
    group = helper.addGroup(monAlg,'default','GLOBAL/BackgroundMon','run')

    # Histogram titles. Some commonly used strings in the histograms defined below.
    bgT = 'Background Word '
    filledT = '(Filled BG) '
    emptyT = '(Empty BG) '
    unpairIsoT = '(UnpairIso BG) '
    unpairNonIsoT = '(UnpairNonIso BG) '
    pixT = 'Pixel SpacePoints '
    sctT = 'SCT SpacePoints '
    hugeMultT = 'Huge Multiplicity '
    lowMultT = 'Low Multiplicity '
    pVertexT = 'Number of Primary Vertices '

    # Background word histograms
    bitSetLabels = ["MBTSTimeDiffHalo","MBTSTimeDiffCol","LArECTimeDiffHalo",
        "LArECTimeDiffCol","PixMultiplicityHuge","PixSPNonEmpty","SCTMultiplicityHuge",
        "SCTSPNonEmpty","CSCTimeDiffHalo","CSCTimeDiffCol","BCMTimeDiffHalo",
        "BCMTimeDiffCol","MuonTimingCol","MuonTimingCosmic","MBTSBeamVeto",
        "BCMBeamVeto","LUCIDBeamVeto","HaloMuonSegment","HaloClusterShape",
        "HaloMuonOneSided","HaloMuonTwoSided","HaloTileClusterPattern","BeamGasPixel",
        "CosmicStandAlone","CosmicStandAloneTight","CosmicCombined",
        "CosmicCombinedTight","BkgdResvBit1","BkgdResvBit2","BkgdResvBit3",
        "BkgdResvBit4","BkgdResvBit5"]
    group.defineHistogram('bitSet', title=bgT, xlabels=bitSetLabels,
                          xbins=len(bitSetLabels),xmin=0,xmax=len(bitSetLabels))
    group.defineHistogram('bitSetFilled', weight='filled',
                          title=bgT+filledT, xlabels=bitSetLabels,
                          xbins=len(bitSetLabels),xmin=0,xmax=len(bitSetLabels))
    group.defineHistogram('bitSetEmpty', weight='empty',
                          title=bgT+emptyT, xlabels=bitSetLabels,
                          xbins=len(bitSetLabels),xmin=0,xmax=len(bitSetLabels))
    group.defineHistogram('bitSetUnpairIso', weight='unpairIso',
                          title=bgT+unpairIsoT, xlabels=bitSetLabels,
                          xbins=len(bitSetLabels),xmin=0,xmax=len(bitSetLabels))
    group.defineHistogram('bitSetUnpairNonIso', weight='unpairNonIso',
                          title=bgT+unpairNonIsoT, xlabels=bitSetLabels,
                          xbins=len(bitSetLabels),xmin=0,xmax=len(bitSetLabels))

    # Pixel spacepoint histograms
    group.defineHistogram('nPixSPs;nPixSPs_high',
                          title=pixT+hugeMultT,
                          xbins=200, xmin=0, xmax=1e6)
    group.defineHistogram('nPixSPs;nPixSPs_high_unpairIso', weight='unpairIso',
                          title=pixT+hugeMultT+unpairIsoT,
                          xbins=200, xmin=0, xmax=1e6)
    group.defineHistogram('nPixSPs;nPixSPs_high_unpairNonIso', weight='unpairNonIso',
                          title=pixT+hugeMultT+unpairNonIsoT,
                          xbins=200, xmin=0, xmax=1e6)

    lowMultiplicityPixGroup = helper.addGroup(monAlg,'lowMultiplicityPixGroup',
                                              'GLOBAL/BackgroundMon','run')
    lowMultiplicityPixGroup.defineHistogram('nPixSPs;nPixSPs_low',
                                            title=pixT+lowMultT,
                                            xbins=200, xmin=0, xmax=5e3)
    lowMultiplicityPixGroup.defineHistogram('nPixSPs;nPixSPs_low_unpairIso',
                                            weight='unpairIso',
                                            title=pixT+lowMultT+unpairIsoT,
                                            xbins=200, xmin=0, xmax=5e3)
    lowMultiplicityPixGroup.defineHistogram('nPixSPs;nPixSPs_low_unpairNonIso',
                                            weight='unpairNonIso',
                                            title=pixT+lowMultT+unpairNonIsoT,
                                            xbins=200, xmin=0, xmax=5e3)

    # SCT spacepoint histograms
    group.defineHistogram('nSctSPs;nSctSPs_high',
                          title=sctT+hugeMultT,
                          xbins=200, xmin=0, xmax=1e6)
    group.defineHistogram('nSctSPs;nSctSPs_high_unpairIso', weight='unpairIso',
                          title=sctT+hugeMultT+unpairIsoT,
                          xbins=200, xmin=0, xmax=1e6)
    group.defineHistogram('nSctSPs;nSctSPs_high_unpairNonIso', weight='unpairNonIso',
                          title=sctT+hugeMultT+unpairNonIsoT,
                          xbins=200, xmin=0, xmax=1e6)

    lowMultiplicitySctGroup = helper.addGroup(monAlg,'lowMultiplicitySctGroup',
                                              'GLOBAL/BackgroundMon','run')
    lowMultiplicitySctGroup.defineHistogram('nSctSPs;nSctSPs_low',
                                            title=sctT+lowMultT,
                                            xbins=200, xmin=0, xmax=1e4)
    lowMultiplicitySctGroup.defineHistogram('nSctSPs;nSctSPs_low_unpairIso',
                                            weight='unpairIso',
                                            title=sctT+lowMultT+unpairIsoT,
                                            xbins=200, xmin=0, xmax=1e4)
    lowMultiplicitySctGroup.defineHistogram('nSctSPs;nSctSPs_low_unpairNonIso',
                                            weight='unpairNonIso',
                                            title=sctT+lowMultT+unpairNonIsoT,
                                            xbins=200, xmin=0, xmax=1e4)

    # LAr histograms
    group.defineHistogram('LArECTimeDiff',
                          title='LAr EC collision time difference;<t_{A}>-<t_{C}> [ns]',
                          xbins=100, xmin=-50, xmax=50)

    # MBTS histograms
    group.defineHistogram('MBTStimeDiff', title='MBTS collision time difference',
                          xbins=100, xmin=-50, xmax=50)
    group.defineHistogram('MBTSvetoHits',
                          title="Total MBTS Hits Setting Veto Bit;Number of MBTS 'veto' hits",
                          xbins=40, xmin=0, xmax=40)

    # Lucid histograms
    group.defineHistogram('LUCIDHits', title='Total Lucid Hits',
                          xbins=64, xmin=0, xmax=32)

    # Beam Background Identification Method
    if monAlg.Muons:
        group.defineHistogram('muonSegmentX,muonSegmentY', type='TH2F',
                              title='Position of the muon segments',
                              xbins=100, xmin=-4, xmax=4,
                              ybins=100, ymin=-4, ymax=4)

        # Distributions of beam background calorimeter clusters
        group.defineHistogram('clusterEnergy', title='Cluster Energy',
                              xbins=150, xmin=0, xmax=1500)
        group.defineHistogram('clusterEta,clusterPhi', type='TH2F', title='Cluster Eta-Phi',
                              xbins=50, xmin=-3, xmax=3, ybins=50, ymin=-3.15, ymax=3.15)
        group.defineHistogram('clusterEta,clusterTime',type='TH2F', title='Cluster Eta-Time',
                              xbins=50, xmin=-3, xmax=3, ybins=50, ymin=-25, ymax=25)

        # Index of the jets tagged by the two-sided method
        group.defineHistogram('jetIndex', title='Fake Jet Index',
                              xbins=10, xmin=0, xmax=10)

        # Distributions in events with a two-sided-method tagged leading jet
        group.defineHistogram('leadingJetPt', title='Fake jet pT',
                              xbins=150, xmin=0, xmax=1500)
        group.defineHistogram('leadingJetEta', title='Fake jet Eta',
                              xbins=50, xmin=-3, xmax=3)
        group.defineHistogram('leadingJetPhi', title='Fake jet Phi',
                              xbins=50, xmin=-3.15, xmax=3.15)
        group.defineHistogram('leadingJetTime', title='Fake jet Time',
                              xbins=120, xmin=-30, xmax=30)
        group.defineHistogram('leadingJetChf', title='Fake jet EM Charge Fraction',
                              xbins=100, xmin=0, xmax=1)
        group.defineHistogram('leadingJetEta,leadingJetTime', type='TH2F', title='Fake jet Eta-Time',
                              xbins=50, xmin=-3, xmax=3, ybins=50, ymin=-25, ymax=25)
        group.defineHistogram('leadingJetEMF,leadingJetCHF', type='TH2F', title='Fake jet EMF-CHF',
                              xbins=50, xmin=0, xmax=1, ybins=50, ymin=0, ymax=1)

        # beam background muon rate per BCID in A->C and C->A direction
        group.defineHistogram('bicd;bcid_AC', weight='isAC', title='Rate per BCID in A->C Direction',
                              xbins=3564, xmin=0, xmax=3564)
        group.defineHistogram('bcid;bcid_CA', weight='isBC', title='Rate per BCID in C->A Direction',
                              xbins=3564, xmin=0, xmax=3564)
        group.defineHistogram('bcid;bcid_ACOneSided', weight='isACOneSided',
                              title='Rate per BCID in A->C Direction (one- and two-sided)',
                              xbins=3564, xmin=0, xmax=3564)
        group.defineHistogram('bcid;bcid_CAOneSided', weight='isCAOneSided',
                              title='Rate per BCID in C->A Direction (one- and two-sided)',
                              xbins=3564, xmin=0, xmax=3564)

    group.defineHistogram('nVertex', title='Number of Primary Vertices;N_{PV}',
                          xbins=50, xmin=0, xmax=50)
    group.defineHistogram('nVertex;nVertex_unpairIso', title=pVertexT+unpairIsoT,
                          weight='unpairIso', xbins=50, xmin=0, xmax=50)
    group.defineHistogram('nVertex;nVertex_unpairNonIso', title=pVertexT+unpairNonIsoT,
                          weight='unpairNonIso', xbins=50, xmin=0, xmax=50)

    acc = helper.result()

    # need background objects
    if not isOld:
        from RecBackgroundAlgs.BackgroundAlgsConfig import BackgroundAlgsCfg
        acc.merge(BackgroundAlgsCfg(flags))

    return acc
